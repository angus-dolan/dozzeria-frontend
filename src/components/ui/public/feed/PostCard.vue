<template>
  <div @click="clickPost()" class="select-none cursor-pointer bg-white border border-gray-200 dark:border-transparent dark:bg-ui-dark-300 mb-3" style="border-radius: 30px">
    <!-- author -->
    <div class="flex pt-4 px-4">
      <avatar 
        :name="'Angus Dolan'" 
        :size="48"  
        :src="'dp.jpg'"
      />
      <div class="flex items-stretch pl-3">
        <span class="place-self-center">
          <h5 class="leading-none text-base font-medium">Angus Dolan</h5>
          <p class="text-dark-low-emphasis dark:text-light-medium-emphasis text-sm">@angus</p>
        </span>
      </div>
    </div>
    <!-- image -->
    <div class='my-3 mx-1 bg-cover h-80 bg-center' style='border-radius: 15px; background-image: url("https://www.pizzacappuccino.com/wp-content/uploads/2019/02/Neapolitan-Pizza-Recipe.jpg")'></div>
    <!-- like button -->
    <div @click.stop="likePost()" class="w-16 px-4 pt-2 pb-2">
      <!-- heroicons:heart -->
      <svg v-if="!postLiked" xmlns="http://www.w3.org/2000/svg" class="pointer-events-none text-dark-high-emphasis dark:text-light-high-emphasis h-7 w-7" fill="none" viewBox="0 0 24 24" stroke="currentColor">
        <path stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z" />
      </svg>
      <svg v-else xmlns="http://www.w3.org/2000/svg" class="pointer-events-none text-red-500 h-7 w-7" fill="currentColor" viewBox="0 0 24 24" stroke="none">
        <path stroke-linecap="round" stroke-linejoin="round" stroke-width="0" d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z" />
      </svg>
    </div>
    <!-- meta -->
    <div class="px-4 pb-4">
      <h5 class="leading-6 font-medium">Classic Neapolitan Margherita</h5>
      <p class="text-sm text-dark-low-emphasis dark:text-light-medium-emphasis font-medium">Wed, 26 January 2021</p>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      postLiked: false,
      delay: 700,
      clicks: 0,
      timer: null
    }
  },
  methods: {
    clickPost() {
      this.clicks++;
      if (this.clicks === 1) {
        this.timer = setTimeout( () => {
          // Execute single click 
          this.clicks = 0
          this.pushToPost()
        }, 
          this.delay
        )
      } else {
          // Execute double click 
          clearTimeout(this.timer)
          this.clicks = 0
          this.likePost()
      }         
    },
    pushToPost() {
      this.$router.push({ name: 'post' })
    },
    likePost() {
      if (!this.postLiked) {
        // Like post
        this.postLiked = true
      } else if (this.postLiked) {
        // Unlike post
        this.postLiked = false
      }
    }
  }
}
</script>
