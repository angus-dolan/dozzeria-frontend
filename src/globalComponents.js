import Vue from 'vue'

/*
  [SHARED]
  - Components used only in public app
*/
Vue.component('public-layout', 
  () => import('@/components/layout/public/AppLayout')
)

/*
  [ADMIN APP]
  - Components used only in admin app
*/
Vue.component('admin-layout', 
  () => import('@/components/layout/admin/AppLayout')
)

/*
  [COMMON]
  - Components shared between admin and public
*/
Vue.component('avatar', 
  () => import('@/components/common/Avatar')
)