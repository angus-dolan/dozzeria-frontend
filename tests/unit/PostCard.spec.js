import { mount } from '@vue/test-utils'

import PostCard from '@/components/ui/public/feed/PostCard.vue'
import '@/globalComponents.js'

const factory = (propsData) => {
  return mount(PostCard, {
    propsData: {
      postImg: 'pizza.jpg',
      postAuthor: 'Angus Dolan',
      postUsername: 'angus',
      title: 'Classic Neapolitan Margherita',
      date: 'Wed, 26 January 2021',
      ...propsData
    }
  })
}

describe('PostCard.vue', () => {

  // displays author name and username
  it ('displays author name and username', () => {
    const wrapper = factory()

    expect(wrapper.find('.author-name').text()).toBe('Angus Dolan')
    expect(wrapper.find('.author-username').text()).toBe('@angus') 
  })

  // like button works as expected
  it ('like button works correctly', () => {
    const wrapper = factory()
    
    wrapper.setData({ postLiked: false })
    const button = wrapper.find('.likebtn')

    // Test liking button
    button.trigger('click')
    expect(wrapper.vm.postLiked).toBe(true)
    // Test unliking button
    button.trigger('click')
    expect(wrapper.vm.postLiked).toBe(false)
  }) 

  it ('double clicking likes a post', () => {
    const wrapper = factory()

    wrapper.setData({ postLiked: false })

    const card = wrapper.find('.postcard')
    expect(wrapper.vm.postLiked).toBe(false)

    // simulate double click
    card.trigger('click')
    card.trigger('click')
  
    // check post is now liked
    expect(wrapper.vm.postLiked).toBe(true)
  })

  // displays post title and date
  it ('displays post title and date', () => {
    const wrapper = factory()

    expect(wrapper.find('.post-title').text()).toBe('Classic Neapolitan Margherita')
    expect(wrapper.find('.post-date').text()).toBe('Wed, 26 January 2021') 
  })

  // regression test
  it ('renders correctly', () => {
    const wrapper = factory()
    expect(wrapper.element).toMatchSnapshot()
  })

})