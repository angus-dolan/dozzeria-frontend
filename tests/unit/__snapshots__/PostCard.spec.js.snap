// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard.vue renders correctly 1`] = `
<div
  class="postcard select-none cursor-pointer bg-white border border-gray-200 dark:border-transparent dark:bg-ui-dark-300 mb-3"
  style="border-radius: 30px;"
>
  <div
    class="flex pt-4 px-4"
  >
    <div>
      <div
        class="rounded-full"
        style="width: 48px; height: 48px;"
      >
        <img
          alt="Image not found :("
          class="w-full h-full"
          src=""
          style="border-radius: inherit;"
        />
      </div>
    </div>
     
    <div
      class="flex items-stretch pl-3"
    >
      <span
        class="place-self-center"
      >
        <h5
          class="author-name leading-none text-base font-medium"
        >
          Angus Dolan
        </h5>
         
        <p
          class="author-username text-dark-low-emphasis dark:text-light-medium-emphasis text-sm"
        >
          @angus
        </p>
      </span>
    </div>
  </div>
   
  <div
    class="postcard__img rounded-xl my-3 mx-1 bg-cover h-80 bg-center"
    style="background-image: url();"
  />
   
  <div
    class="likebtn w-16 px-4 pt-2 pb-2"
  >
    <svg
      class="likebtn__unliked pointer-events-none text-dark-high-emphasis dark:text-light-high-emphasis h-7 w-7"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
      />
    </svg>
  </div>
   
  <div
    class="px-4 pb-4"
  >
    <h5
      class="post-title leading-6 font-medium"
    >
      Classic Neapolitan Margherita
    </h5>
     
    <p
      class="post-date text-sm text-dark-low-emphasis dark:text-light-medium-emphasis font-medium"
    >
      Wed, 26 January 2021
    </p>
  </div>
</div>
`;
